src/geometry/ArborX_Point.hpp

```diff
@@ -8,72 +8,55 @@
  *                                                                          *
  * SPDX-License-Identifier: BSD-3-Clause                                    *
  ****************************************************************************/
-
-#ifndef ARBORX_DETAILS_POINT_HPP
-#define ARBORX_DETAILS_POINT_HPP
+#ifndef ARBORX_POINT_HPP
+#define ARBORX_POINT_HPP
 
 #include <ArborX_GeometryTraits.hpp>
 
 #include <Kokkos_Macros.hpp>
 
-#include <utility>
-
 namespace ArborX
 {
-class Point
-{
-private:
-  struct Data
-  {
-    float coords[3];
-  } _data = {};
 
-  struct Abomination
-  {
-    double xyz[3];
-  };
+template <int DIM, class Coordinate = float>
+struct Point
+{
+  static_assert(DIM > 0);
 
-public:
-  KOKKOS_DEFAULTED_FUNCTION
-  constexpr Point() noexcept = default;
+  KOKKOS_FUNCTION
+  constexpr auto &operator[](unsigned int i) { return _coords[i]; }
 
-  KOKKOS_INLINE_FUNCTION
-  constexpr Point(Abomination data)
-      : Point(static_cast<float>(data.xyz[0]), static_cast<float>(data.xyz[1]),
-              static_cast<float>(data.xyz[2]))
-  {}
+  KOKKOS_FUNCTION
+  constexpr auto const &operator[](unsigned int i) const { return _coords[i]; }
 
-  KOKKOS_INLINE_FUNCTION
-  constexpr Point(float x, float y, float z)
-      : _data{{x, y, z}}
-  {}
+  // Initialization is needed to be able to use Point in constexpr
+  // TODO: do we want to actually want to zero initialize it? Seems like
+  // unnecessary work.
+  Coordinate _coords[DIM] = {};
+};
 
-  KOKKOS_INLINE_FUNCTION
-  constexpr float &operator[](unsigned int i) { return _data.coords[i]; }
+template <typename... T>
+KOKKOS_FUNCTION Point(T...)
+    -> Point<sizeof...(T), std::conditional_t<
+                               (... || std::is_same_v<std::decay_t<T>, double>),
+                               double, float>>;
```

Comment:
```suggestion
#if KOKKOS_VERSION >= 40400
KOKKOS_DEDUCTION_GUIDE
#else
KOKKOS_FUNCTION 
#endif
Point(T...)
    -> Point<sizeof...(T), std::conditional_t<
                               (... || std::is_same_v<std::decay_t<T>, double>),
                               double, float>>;
```

Reply:
Thank you. Not touching it here, but will make it a part of #1145.